File: dotnet/src/Microsoft.AutoGen/Core/IHandle.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// IHandle.cs

using Google.Protobuf;
using Microsoft.AutoGen.Contracts;

namespace Microsoft.AutoGen.Core;

/// <summary>
/// Defines a handler interface for processing items of type <typeparamref name="T"/>.
/// </summary>
/// <typeparam name="T">The type of item to be handled, which must implement <see cref="IMessage"/>.</typeparam>
public interface IHandle<in T>
{
    /// <summary>
    /// Handles the specified item asynchronously.
    /// </summary>
    /// <param name="item">The item to be handled.</param>
    /// <returns>A task that represents the asynchronous operation.</returns>
    ValueTask HandleAsync(T item, MessageContext messageContext);
}

public interface IHandle<in InT, OutT>
{
    /// <summary>
    /// Handles the specified item asynchronously.
    /// </summary>
    /// <param name="item">The item to be handled.</param>
    /// <returns>A task that represents the asynchronous operation.</returns>
    ValueTask<OutT> HandleAsync(InT item, MessageContext messageContext);
}

File: typescript/src/contracts/IHandle.ts
import { MessageContext } from "./MessageContext";

/**
 * Defines a handler interface for processing items of type TMessage.
 * @template TMessage The type of message to be handled.
 */
export interface IHandle<TMessage> {
  /**
   * Handles the specified message asynchronously.
   * @param message The message to be handled.
   * @param context The context information for the message being handled.
   * @returns A promise that resolves to the result of handling the message.
   */
  handleAsync(message: TMessage, context: MessageContext): Promise<unknown>;
}